src/analyses/relationPriv.ml

```diff
@@ -45,340 +42,344 @@ module type S =
+    let read_global ask getg st g x = st.apr
+    let write_global ?(invariant=false) ask getg sideg st g x = st
 
-  let lock ask getg st m = st
-  let unlock ask getg sideg st m = st
+    let lock ask getg st m = st
+    let unlock ask getg sideg st m = st
 
-  let thread_join ask getg exp st = st
-  let thread_return ask getg sideg tid st = st
+    let thread_join ask getg exp st = st
+    let thread_return ask getg sideg tid st = st
 
-  let sync ask getg sideg st reason = st
+    let sync ask getg sideg st reason = st
 
-  let enter_multithreaded ask getg sideg st = st
-  let threadenter ask getg st = st
+    let enter_multithreaded ask getg sideg st = st
+    let threadenter ask getg st = st
 
-  let init () = ()
-  let finalize () = ()
-end
+    let init () = ()
+    let finalize () = ()
+  end
 
 module type ProtectionBasedPrivParam =
 sig
   (** Whether to be path-sensitive w.r.t. locally written protected globals that have been continuously protected since writing. *)
   val path_sensitive: bool
 end
 
-(** Protection-Based Reading. *)
-module ProtectionBasedPriv (Param: ProtectionBasedPrivParam): S = functor (AD: ApronDomain.S2) ->
-struct
-  include ConfCheck.RequireMutexActivatedInit
-  open Protection
-
-  (** Locally must-written protected globals that have been continuously protected since writing. *)
-  module P =
-  struct
-    include MustVars
-    let name () = "P"
-  end
+(** Protection-Based Reading. **)
 
-  (** Locally may-written protected globals that have been continuously protected since writing. *)
-  (* TODO: is this right? *)
-  module W =
+module ProtectionBasedPriv (Param: ProtectionBasedPrivParam): S =
+  functor (RD: RelationDomain.RD) ->
   struct
-    include MayVars
-    let name () = "W"
-  end
-
-  module D = Lattice.Prod (P) (W)
-  module G = AD
-  module V = Printable.UnitConf (struct let name = "global" end)
-
-  type apron_components_t = ApronComponents (AD) (D).t
-
-  module VM =
-  struct
-    type t =
-      | Local of varinfo
-      | Unprot of varinfo
-      | Prot of varinfo
-
-    let var_name = function
-      | Local g -> g.vname
-      | Unprot g -> g.vname ^ "#unprot"
-      | Prot g -> g.vname ^ "#prot"
-  end
-  module AV =
-  struct
-    include ApronDomain.VarMetadataTbl (VM)
-    open VM
-
-    let local g = make_var (Local g)
-    let unprot g = make_var (Unprot g)
-    let prot g = make_var (Prot g)
-  end
+    include ConfCheck.RequireMutexActivatedInit
+    open Protection
+
+    (** Locally must-written protected globals that have been continuously protected since writing. *)
+    module P =
+    struct
+      include MustVars
+      let name () = "P"
+    end
+
+    (** Locally may-written protected globals that have been continuously protected since writing. *)
+    (* TODO: is this right? *)
+    module W =
+    struct
+      include MayVars
+      let name () = "W"
+    end
+
+    module D = Lattice.Prod (P) (W)
+    module G = RD.D2
```

Comment:
The fact that instead of `AD` one has to now use `RD.D2` everywhere is very annoying. The domain module contains a domain module in itself?

Reply:
Makes sense. `D2` is now directly included inside `RD`